src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +141,94 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
             };
         }
 
-        private async Task<bool> RestoreCoreAsync(ProjectRestoreInfo restoreInfo)
+        private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
         {
+            ProjectRestoreInfo? restoreInfo = value.RestoreInfo;
+
+            // Prevent extra restore under some conditions.
+            if (IsProjectConfigurationVersionOutOfDate(value.ConfiguredInputs))
+            {
+                return false;
+            }
+
             // Restore service always does work regardless of whether the value we pass 
             // them to actually contains changes, only nominate if there are any.
-            byte[] hash = RestoreHasher.CalculateHash(restoreInfo);
+            byte[] hash = RestoreHasher.CalculateHash(restoreInfo!);
 
             if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
+            {
+                SaveNominatedConfiguredVersions(value.ConfiguredInputs);
                 return true;
+            }
 
             _latestHash = hash;
 
+            _nugetRestoreStarted = true;
             JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
             {
-                return NominateForRestoreAsync(restoreInfo, _projectAsynchronousTasksService.UnloadCancellationToken);
+                return NominateForRestoreAsync(restoreInfo!, _projectAsynchronousTasksService.UnloadCancellationToken);
             });
 
+            SaveNominatedConfiguredVersions(value.ConfiguredInputs);
+
             _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
                                                                ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
                                                                registerFaultHandler: true);
 
             // Prevent overlap until Restore completes
             bool success = await joinableTask;
 
-            HintProjectDependentFile(restoreInfo);
+            _nugetRestoreStarted = false;
```

Comment:
this _nugetRestoreStarted need be done in the same lock region as SaveNominatedConfiguredVersions, otherwise, we leave a race condition in the code.